File: 02_25/PLP/tai_6/Main.java
import java.util.ArrayList;
import java.util.List;

abstract class Pessoa {
    /*
     * Classe Pai abstrata - não pode ser instanciada diretamente.
     */
    public String nome;
    public int idade;

    // Construtor (equivalente ao __init__)
    public Pessoa(String nome, int idade) {
        this.nome = nome;
        this.idade = idade;
    }

    // Define método como abstrato; as subclasses são obrigadas a implementar.
    public abstract String realizarAtividade();

    public void apresentar() {
        System.out.printf("Olá, meu nome é %s e tenho %d anos.%n", this.nome, this.idade);
    }
}

class Professor extends Pessoa {
    private String disciplina;

    public Professor(String nome, int idade, String disciplina) {
        // chama o construtor da classe pai (super().__init__)
        super(nome, idade);
        this.disciplina = disciplina;
    }

    // Sobrescrita (override) da abstrata. Implementação específica para Professor.
    @Override
    public String realizarAtividade() {
        return String.format("%s está lecionando a matéria : %s.", this.nome, this.disciplina);
    }
}

class Aluno extends Pessoa {
    private String matricula;

    public Aluno(String nome, int idade, String matricula) {
        // chama o construtor da classe pai
        super(nome, idade);
        this.matricula = matricula;
    }

    // Sobrescrita (override) da abstrata. Implementação específica para Aluno.
    @Override
    public String realizarAtividade() {
        return String.format("%s (Matrícula : %s) está estudando para prova.", this.nome, this.matricula);
    }
}

public class Main {
    public static void main(String[] args) {
        System.out.println("-----    Demonstração de Herança e Classe Abstrata    -----\n");

        Professor p1 = new Professor("André", 35, "Paradigmas de Linguagens de Programação");
        Aluno p2 = new Aluno("Nathã", 18, "1250109474");
        // Pessoa p3 = new Pessoa("João", 25); // Dará ERRO ao tentar instanciar a classe abstrata.

        // Lista para exemplo de polimorfismo
        List<Pessoa> lista = new ArrayList<>();
        lista.add(p1);
        lista.add(p2);

        for (Pessoa pessoa : lista) {
            // Chamada usa método da classe Pessoa, mas é executado na classe real Professor/Aluno
            String atividade = pessoa.realizarAtividade();
            pessoa.apresentar(); // método herdado base

            System.out.println("Atividade Principal : " + atividade); // método polimórfico
            System.out.println("-".repeat(30));
            System.out.println();
        }
    }
}

File: 02_25/PLP/tai_6/main.py
from abc import ABC, abstractmethod # importação biblioteca de abstração
 
class Pessoa(ABC):
    '''
    Herda ABC, o que a torna abstrata.
    Classe Pai abstrata - não pode ser instanciada diretamente. 
    '''
    def __init__(self, nome, idade):
        self.nome = nome
        self.idade = idade

    @abstractmethod # define método como abstrato, as subclasses são obrigadas a implementar este método.
    def realizar_atividade(self):
        pass

    def apresentar(self):
        print(f'Olá, meu nome é {self.nome} e tenho {self.idade} anos.')

class Professor(Pessoa):
    def __init__(self, nome, idade, disciplina):
        # chama o construtor da classe pai
        super().__init__(nome, idade)
        self.disciplina = disciplina

    # sobrescrita (override) da abstrata. implementação específica para classe Professor.
    def realizar_atividade(self):
        return f'{self.nome} está lecionando a matéria : {self.disciplina}.'
    
class Aluno(Pessoa):
    def __init__(self, nome, idade, matricula):
        # chama o construtor da classe pai
        super().__init__(nome, idade)
        self.matricula = matricula

    # sobrescrita (override) da abstrata. implementação específica para classe Aluno.
    def realizar_atividade(self):
        return f'{self.nome} (Matrícula : {self.matricula}) está estudando para prova.'
    
if __name__ == '__main__':
    print("-----    Demonstração de Herança e Classe Abstrata   -----\n")
    p1 = Professor('André', 35, 'Paradigmas de Linguagens de Programação')
    p2 = Aluno('Nathã', 18, '1250109474')
    # p3 = Pessoa('João', 25) Dará ERROR ao tentar instanciar a class abstrata.

    # lista para exemplo de polimorfismo
    lista = [p1, p2]


    for pessoa in lista:
        # chamada usa método da classe Pessoa, mas é executado na classe real Professor/Aluno
        atividade = pessoa.realizar_atividade()
        pessoa.apresentar() # método herdado base

        print(f'Atividade Principal : {atividade}') # método polimórfico
        print('-' * 30)
        print()